LayoutTests/inspector/console/setConsoleClearAPIEnabled.html

```diff
@@ -0,0 +1,111 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="../../http/tests/inspector/resources/inspector-test.js"></script>
+<script>
+function test()
+{
+    let suite = InspectorTest.createAsyncSuite("Console.setConsoleClearAPIEnabled");
+
+    suite.addTestCase({
+        name: "Console.setConsoleClearAPIEnabled.API.Disabled",
+        description: "Check that console.clear() doesn't clear the console when disallowed.",
+        async test() {
+            await ConsoleAgent.setConsoleClearAPIEnabled(false);
+
+            let cleared = false;
+            let clearedListener = WI.consoleManager.addEventListener(WI.ConsoleManager.Event.Cleared, (event) => {
+                cleared = true;
+            });
+
+            let messageAdded = false;
+            let messageAddedListener = WI.consoleManager.addEventListener(WI.ConsoleManager.Event.MessageAdded, (event) => {
+                messageAdded = true;
+            });
+
+            await InspectorTest.evaluateInPage(`console.clear()`);
+
+            InspectorTest.expectFalse(cleared, "Should not clear console.");
+            InspectorTest.expectFalse(messageAdded, "Should not send a message.");
+
+            WI.consoleManager.removeEventListener(WI.ConsoleManager.Event.Cleared, clearedListener);
+            WI.consoleManager.removeEventListener(WI.ConsoleManager.Event.MessageAdded, messageAddedListener);
+        },
+    });
+
+    suite.addTestCase({
+        name: "Console.setConsoleClearAPIEnabled.API.Enabled",
+        description: "Check that console.clear() does clear the console when allowed.",
+        async test() {
+            await ConsoleAgent.setConsoleClearAPIEnabled(true);
+
+            let cleared = false;
+            let clearedListener = WI.consoleManager.addEventListener(WI.ConsoleManager.Event.Cleared, (event) => {
+                cleared = true;
+            });
+
+            let messageAdded = false;
+            let messageAddedListener = WI.consoleManager.addEventListener(WI.ConsoleManager.Event.MessageAdded, (event) => {
+                messageAdded = true;
+            });
+
+            await InspectorTest.evaluateInPage(`console.clear()`);
+
+            InspectorTest.expectTrue(cleared, "Should clear console.");
+            InspectorTest.expectFalse(messageAdded, "Should not send a message.");
```

Comment:
Same question here: what ensures the event listeners did run by the time these assertions are made?

Reply:
FWIW we should never dispatch `WI.ConsoleManager.Event.MessageAdded` when receiving a `Console.messageAdded` as we have a special `Console.messagesCleared` for that reason

Reply:
My question is more about what prevents a race condition in this test now or in the future: the assertions running before the event handlers change the boolean flags?

Reply:
FWIW we have assumptions like this _all_ over Web Inspector code since our protocol (which is backed by XPC) is guaranteed to deliver messages in the order that they're sent

in this case ive added a `console.log()` after the `console.clear()` to ensure that a `Console.messageAdded` comes after a `Console.messageCleared`

Reply:
Fair enough. At first glance this looks vulnerable to race conditions, but I'm satisfied with the explanation about the guaranteed order of messages over the protocol.